types/src/validator_verifier.rs

```diff
@@ -159,54 +166,99 @@ impl ValidatorVerifier {
+        Ok((
+            AggregatedSignature::new(validator_bitmask, Some(aggregated_sig)),
+            aggregated_key,
+        ))
+    }
+
+    pub fn verify_and_generate_aggregated_signature<T: CryptoHash + Serialize>(
+        &self,
+        partial_signatures: &PartialSignatures,
+        message: &T,
+    ) -> Result<AggregatedSignature, VerifyError> {
+        let (aggregated_sig, aggregated_key) =
+            self.generate_aggregated_signature(partial_signatures)?;
+        aggregated_sig
+            .aggregated_signature()
+            .as_ref()
+            .ok_or(VerifyError::FailedToAggregateKey)?
+            .verify(message, &aggregated_key)
+            .expect("Failed to verify the aggregated bls signature");
+        Ok(aggregated_sig)
     }
 
-    /// This function will try batch signature verification and falls back to normal
-    /// iterated verification if batching fails.
-    pub fn batch_verify_aggregated_signatures<T: CryptoHash + Serialize>(
+    pub fn verify_aggregated_signatures<T: CryptoHash + Serialize>(
         &self,
         message: &T,
-        aggregated_signature: &BTreeMap<AccountAddress, bls12381::Signature>,
+        aggregated_signature: &AggregatedSignature,
     ) -> std::result::Result<(), VerifyError> {
         self.check_num_of_signatures(aggregated_signature)?;
-        self.check_voting_power(aggregated_signature.keys())?;
-        let keys_and_signatures: Vec<(bls12381::PublicKey, bls12381::Signature)> =
+        self.check_voting_power(
             aggregated_signature
-                .iter()
-                .flat_map(|(address, signature)| {
-                    let sig = signature.clone();
-                    self.get_public_key(address).map(|pub_key| (pub_key, sig))
-                })
-                .collect();
-        // Fallback is required to identify the source of the problem if batching fails.
-        if bls12381::Signature::batch_verify(message, keys_and_signatures).is_err() {
-            self.verify_aggregated_struct_signature(message, aggregated_signature)?
-        }
+                .get_voter_addresses(&self.get_ordered_account_addresses_iter().collect_vec())
+                .iter(),
+        )?;
+        // Verify the optimistically aggregated signature.
+        let pub_keys_to_agg: Vec<&PublicKey> = aggregated_signature
+            .get_voters_bitmap()
+            .iter()
+            .zip(&self.address_to_validator_info)
+            .filter_map(|(voted, (_address, consensus_info))| {
+                if *voted {
+                    Some(&consensus_info.public_key)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        let aggregated_key =
+            PublicKey::aggregate(pub_keys_to_agg).map_err(|_| VerifyError::FailedToAggregateKey)?;
+
+        aggregated_signature
+            .aggregated_signature()
+            .as_ref()
+            .ok_or(VerifyError::InvalidSignature)?
```

Comment:
we probably should do this None check earlier

Reply:
This is okay because some tests rely on passing empty signature, which is handled by check  `if pub_keys_to_agg.is_empty() {`

Reply:
doing simple check early can provide better defense on invalid sigs

Reply:
this is again needed for tests to pass. We can't move it up before `pub_keys_to_agg.is_empty() {` check.

Reply:
we can move the null check to the beginning now?